src/codecs/openexr.rs

```diff
@@ -0,0 +1,184 @@
+// - [] ImageEncoder
+// - [] GenericImageView?
+// - [] GenericImage?
+// - [] Progress
+
+// # MAYBE SOON
+// - [] ImageDecoderExt::read_rect_with_progress
+// - [] Layers -> Animation?
+
+
+extern crate exr;
+use exr::prelude::*;
+
+use crate::{ImageDecoder, ImageResult, ColorType, Progress, image, ImageError, ImageFormat};
+use std::io::{Write, Seek, BufRead, SeekFrom, Cursor};
+use crate::error::{DecodingError, ImageFormatHint};
+use self::exr::meta::header::Header;
+
+/// An OpenEXR decoder
+#[derive(Debug)]
+pub struct ExrDecoder<R> {
+    source: R,
+    header: exr::meta::header::Header,
+}
+
+
+
+impl<R: BufRead + Seek + Send> ExrDecoder<R> {
+
+    /// Create a decoder. Consumes the first few bytes of the source to extract image dimensions.
+    pub fn read(mut source: R) -> ImageResult<Self> {
+
+        // read meta data, then go back to the start of the file
+        let mut meta_data = {
+            let start_position = source.stream_position()?;
+            let meta: MetaData = exr::meta::MetaData::read_from_buffered(&mut source, true)?;
+            source.seek(SeekFrom::Start(start_position))?;
+            meta
+        };
+
+        let header: Header = meta_data.headers.into_iter()
+            .filter(|header|{
+                let has_rgb = ["R","G","B"].iter().all(
+                    |required| header.channels.list.iter().find(|chan| chan.name.eq(required)).is_some() // TODO eq_lowercase only if exrs supports it
+                );
+
+                !header.deep && has_rgb
+            })
+            .next()
+            .ok_or_else(|| ImageError::Decoding(DecodingError::new(
+                ImageFormatHint::Exact(ImageFormat::Exr),
+                "image does not contain non-deep rgb channels"
+            )))?;
+
+        Ok(Self { source, header })
+    }
+}
+
+
+impl<'a, R: 'a + BufRead + Seek + Send> ImageDecoder<'a> for ExrDecoder<R> {
+    type Reader = Cursor<Vec<u8>>;
+
+    fn dimensions(&self) -> (u32, u32) {
+        let s = self.header.layer_size;
+        (s.width() as u32, s.height() as u32)
+    }
+
+    // TODO fn original_color_type(&self) -> ExtendedColorType {}
+
+    fn color_type(&self) -> ColorType {
+        // TODO adapt to actual exr color type
+        // TODO f32 ColorType::Rgba32
+        ColorType::Rgba16
+    }
+
+    /// Use `read_image` if possible,
+    /// as this method creates a whole new buffer to contain the entire image.
+    fn into_reader(self) -> ImageResult<Self::Reader> {
+        Ok(Cursor::new(image::decoder_to_vec(self)?)) // TODO no vec?
+    }
+
+    fn scanline_bytes(&self) -> u64 {
+        // we cannot always read individual scan lines,
+        // as the tiles or lines in the file could be in random order.
+        // therefore we currently read all lines at once
+        // Todo: respect exr.line_order
+       self.total_bytes()
+    }
+
+    fn read_image(mut self, target: &mut [u8]) -> ImageResult<()> {
+        self.read_image_with_progress(target, |_|{})
```

Comment:
This is the default implementation, so you don't have to duplicate it unless you're planning on making an optimized version later.

Reply:
I know. I implemented it because this should always be true, even if the default implementation changes in the future.